tm2/pkg/std/tx_test.go

```diff
@@ -0,0 +1,280 @@
+	multisigPubKey := multisig.NewPubKeyMultisigThreshold(2, pubKeys)
+	require.Equal(t, len(pubKeys), CountSubKeys(multisigPubKey))
+}
+
+func TestGetSigners(t *testing.T) {
+	// Single signer case
+	addr, _ := crypto.AddressFromBech32("g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5")
+	msgs := []Msg{
+		mockMsg{
+			caller:  addr,
+			msgType: "call",
+		},
+	}
+	tx := NewTx(msgs, Fee{}, []Signature{}, "")
+	require.Equal(t, []crypto.Address{addr}, tx.GetSigners())
+
+	// Duplicate signers case
+	msgs = []Msg{
+		mockMsg{
+			caller:  addr,
+			msgType: "send",
+		},
+		mockMsg{
+			caller:  addr,
+			msgType: "send",
+		},
+	}
+
+	tx = NewTx(msgs, Fee{}, []Signature{}, "")
+	require.Equal(t, []crypto.Address{addr}, tx.GetSigners())
+
+	// Multiple unique signers case
+	addr2, _ := crypto.AddressFromBech32("g1us8428u2a5satrlxzagqqa5m6vmuze025anjlj")
+	msgs = []Msg{
+		mockMsg{
+			caller:  addr,
+			msgType: "call",
+		},
+		mockMsg{
+			caller:  addr2,
+			msgType: "run",
+		},
+	}
+	tx = NewTx(msgs, Fee{}, []Signature{}, "")
+	require.Equal(t, []crypto.Address{addr, addr2}, tx.GetSigners())
+
+	// "no_op" message case
+	msgs = []Msg{
+		mockMsg{
+			caller:  addr,
+			msgType: "no_op",
+		},
+		mockMsg{
+			caller:  addr2,
+			msgType: "call",
+		},
+	}
+	tx = NewTx(msgs, Fee{}, []Signature{}, "")
+	require.Equal(t, []crypto.Address{addr, addr2}, tx.GetSigners())
+}
+
+func TestGetSignBytes(t *testing.T) {
+	msgs := []Msg{}
+	fee := NewFee(1000, Coin{Denom: "atom", Amount: 10})
+	sigs := []Signature{}
+	tx := NewTx(msgs, fee, sigs, "test memo")
+	chainID := "test-chain"
+	accountNumber := uint64(1)
+	sequence := uint64(1)
+
+	signBytes, err := tx.GetSignBytes(chainID, accountNumber, sequence)
+	require.NoError(t, err)
+	require.NotEmpty(t, signBytes)
+}
+
+func TestIsSponsorTx(t *testing.T) {
+	addr1, _ := crypto.AddressFromBech32("g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5")
+	addr2, _ := crypto.AddressFromBech32("g1us8428u2a5satrlxzagqqa5m6vmuze025anjlj")
+
+	tests := []struct {
+		name     string
+		msgs     []Msg
+		expected bool
+	}{
+		{
+			name: "no_op message with different signers",
+			msgs: []Msg{
+				mockMsg{
+					caller:  addr1,
+					msgType: "no_op",
+				},
+				mockMsg{
+					caller:  addr2,
+					msgType: "call",
+				},
+			},
+			expected: true,
+		},
+		{
+			name: "only have no_op message",
+			msgs: []Msg{
+				mockMsg{
+					caller:  addr1,
+					msgType: "no_op",
+				},
+			},
+			expected: false,
+		},
+		{
+			name: "no_op message with same signers",
+			msgs: []Msg{
+				mockMsg{
+					caller:  addr1,
+					msgType: "no_op",
+				},
+				mockMsg{
+					caller:  addr1,
+					msgType: "call",
+				},
+			},
+			expected: false,
+		},
+		{
+			name: "non no_op message",
+			msgs: []Msg{
+				mockMsg{
+					caller:  addr1,
+					msgType: "call",
+				},
+				mockMsg{
+					caller:  addr2,
+					msgType: "send",
+				},
+			},
+			expected: false,
```

Comment:
should be true. sponsor isn't because no_op is the first message, but because the tx contains multiple signers, and we just take the first one as the payer.